ext/spl/spl_array.c

```diff
@@ -502,6 +491,19 @@ static void spl_array_write_dimension_ex(int check_inherited, zend_object *objec
 		return;
 	}
 
+	if (check_inherited && intern->fptr_offset_set) {
+		zval tmp;
+
+		if (!offset) {
+			ZVAL_NULL(&tmp);
+			offset = &tmp;
+		}
+		zend_call_method_with_2_params(object, object->ce, &intern->fptr_offset_set, "offsetSet", NULL, offset, value);
+		return;
+	}
```

Comment:
This should be done in the `!offset` branch as well: https://3v4l.org/AYJjF

Reply:
Not sure if I understand what you are asking me to do here?

Reply:
In the `if (!offset) {` branch above, you need to call `offsetSet()` if there is such method on the object. https://3v4l.org/AYJjF shows that `offsetSet()` used to be called during `$arrayObject[] = ...;`

Reply:
With the code at https://3v4l.org/lO4Tf, this is the output on master:
```
string(12) "A::offsetSet"
string(12) "B::offsetSet"
```
and this is the output on this branch:
```
string(12) "A::offsetSet"
```

Is this an expected behavior change? If not, I think that we should move the `if (check_inherited && intern->fptr_offset_set) {` block above `if (!offset) {`

Reply:
Huuuum I think it is slightly problematic that an object that implements ArrayAccess supports appending as it is indistinguishable from ``$o[null]`` but this is something I should add tests for in #12723 

But I do agree that ArrayObject subclassing and ArrayAccess implementations should behave identically. Thanks for spotting this, I'm amending this and adding a test :)